### ISSUE TITLE
proposal: allow skipping templating engine for manifests

### ISSUE THREAD

**Issue by invidian** [helm/helm#8497]
It would be nice to have an ability to skip rendering engine for certain
manifests, when building the chart object programmatically.

This would allows to completely skip templating part of helm, while still be
able to use hooks, versioning, upgrades etc. functionality.

Assembling charts programmatically is useful, for example when user wants to use
different templating mechanism for generating the manifests, e.g. kustomize, but still use
helm for managing the installation.

While for many charts it is fine to be processed by templating engine
multiple times, for others, it is problematic.

The example here are charts, which use similar templating syntax in ConfigMap
objects, so they have to escape {{ }} templating syntax, which is
already non-trivial with helm. The example of such chart is
prometheus-operator, which brings alerting rules, which also use {{ }}
syntax.

Having an option to skip Helm's templating engine allows to skip painful process of escaping such syntax.

Currently in Helm, only `crds` directory is not being templated, however objects from there are not managed
by Helm.

I'm mainly interested in Go functionality, however this feature could also be exposed to the CLI frontend.

**Comment by jdolitsky** [comment]
@invidian when you say ability to skip templating, do you mean upon installation? 

I'm wondering if what you're trying to achieve can be done with post-rendering? please see here: https://helm.sh/docs/topics/advanced/#post-rendering

One of the primary use cases of this was to work with Kustomize. You could essentially not put any templating in your .yaml files at all and use some other process to render manifests appropriately.

**Comment by invidian** [comment]
Hmm, thanks for the idea @jdolitsky. It seems using post-rendering for that should work, though the API is not that nice, as manifests are no longer part of the chart, but comes sort of from the outside. I'll give it a try.

**Comment by mattfarina** [comment]
I think it's important to look at this from the vantage point of chart creators and chart consumers. They are often two different people at two different organizations.

Post rendering is a feature for chart consumers. It provides a means for them to make changes to a chart the chart creator may not have made possible. I don't think it's a good idea for a chart creator to require a chart consumer to use a post render hook. This will create problems. For example, Helm works on windows, mac, and linux. The same script for Linux may not work on windows. And, sometimes scripts for linux don't work on mac (which is BSD based).

Not templating a complete file causes other problems. For example, is the way a name is set. How does that get set to something relating to what a chart consumer chooses? Today naming patterns are used but those use template.

There is a way to handle this today using the Helm chart [file handling](https://helm.sh/docs/chart_template_guide/accessing_files/).

First, define you custom configuration in it's own file. I created a `foo.yaml` file at the root of a chart with the contents...

```
yar: {{ "wee" }}
```

In the _templates_ directory create something that needs to have it. Something like a ConfigMap or Secret. In it you want to set the contents of the file as a [multiline string](https://yaml-multiline.info/). Here's a simple example...

```
grr: |-
{{- .Files.Get "foo.yaml" | nindent 2}}
```

Helm will render the output of this as...

```
grr: |-
  yar: {{ "wee" }}
```

You'll need to put this into a proper ConfigMap or Secret format and then consume that on the other side (e.g., as a file mounted in the container or something). But, it is a way to pass a template around today.

**Comment by invidian** [comment]
Maybe to add a reproducible case for my issue, here are the steps:
```sh
docker run -it --entrypoint /bin/sh alpine/helm:3.2.4 -c "helm create foo && helm template foo ./foo && helm repo add stable https://kubernetes-charts.storage.googleapis.com/ && helm repo update && helm template foo stable/prometheus-operator > foo/templates/bar.yaml && helm template foo ./foo"
```

If I could instead render `prometheus-operator` manifests into `foo/manfiests`, then everything would be fine.

The same applies for Go API.

**Comment by mattfarina** [comment]
@invidian I was able to replicate your situation. I provided a method to work around it without any changes to Helm being required.

The changes need to happen within a chart. It's entirely possible to pass content to Kubernetes that's part templated and processed and part that's templated and not processed. It just needs to be broken into multiple files within the chart and use the file handling.

**Comment by invidian** [comment]
>@invidian I was able to replicate your situation. I provided a method to work around it without any changes to Helm being required.

>The changes need to happen within a chart. It's entirely possible to pass content to Kubernetes that's part templated and processed and part that's templated and not processed. It just needs to be broken into multiple files within the chart and use the file handling.

Unfortunately, the method you provided does not solve the issue, as Helm still interprets the template syntax twice in this case. To reproduce what you suggested:
```sh
helm create foo
echo "foo: {{ .wee }}" > foo/foo.yaml
echo -e "grr: |-\n{{- .Files.Get \"foo.yaml\" | nindent 2}}" >  foo/templates/foo.yaml
helm create baz
helm template foo ./foo/ > baz/templates/foo.yaml
helm template baz baz/
```

It ends up with following output (irrelevant bits skipped):
```
---
# Source: baz/templates/foo.yaml
# Source: foo/templates/foo.yaml
grr: |-
  foo:
```

Instead of expected:
```
# Source: baz/templates/foo.yaml
# Source: foo/templates/foo.yaml
grr: |-
  foo: {{ .wee }}
```

**Comment by mattfarina** [comment]
@invidian here's an example... https://github.com/mattfarina/example-chart-with-go-templates-rendered

If, from the root of the directory, you run... `helm template foo .` you'll see what I'm talking about. This works with `helm install` and the other commands.

**Comment by invidian** [comment]
@mattfarina thanks for your example, but see, that in my case, I run `helm template` twice, on 2 different charts and this is crucial to my process, as I am generating the manifests using chart X and then add those to chart Y + other manifests. The trick with post-renderer works fine for this process, though it's not the most elegant.

**Comment by invidian** [comment]
Not stale.

**Comment by invidian** [comment]
Not stale, but I didn't have time to get back to it :disappointed:

**Comment by invidian** [comment]
Not stale.

**Comment by Nitive** [comment]
For everyone who interested — we achieved skiping templating part with a bunch of hacks and now we use TypeScript for writing templates and it's been a great experience. We packed our solution into plugin [helm-deno](https://github.com/CSSSR/helm-deno#current-limitationshttps://github.com/CSSSR/helm-deno) and open-sourced the code. It has [some limitations](https://github.com/CSSSR/helm-deno#current-limitations) but it works and we use it in production for more than a 7 months. We don't actively implement features we don't need in our company but we are open for contributions
